### ISSUE TITLE
rope range for linear models

### ISSUE THREAD

**Issue by mattansb** [easystats/bayestestR#364]
For logistic regression, the rope range is a function of 3/sqrt(pi), as this is the sd of the logistic distribution (on the latent scale).

However, this is the _conditional_ sd, which is akin to sigma in a linear model.
But in a linear model, the rope range if a function of the marginal sd, no?

This seems odd to me. Have I missed something?

**Comment by DominiqueMakowski** [comment]
mmh I think the rationale we followed back then was quite simple: finding the equivalent of 0.1 * SD, and the most straightforward way was using the d -> log odds conversion formula... but indeed, there might be more appropriate ways of tackling this, what would you recommend?

**Comment by mattansb** [comment]
I thought this was based on Kruschke's book, no?

**Comment by DominiqueMakowski** [comment]
afaik he only gave recommendation for linear models and from there we derived the logistic one

**Comment by strengejacke** [comment]
> afaik he only gave recommendation for linear models and from there we derived the logistic one

No, see

![image](https://user-images.githubusercontent.com/26301769/105631741-3f75a000-5e50-11eb-95f2-fdb84fbf68cf.png)

(from Supplement https://doi.org/10.1177/2515245918771304)

**Comment by strengejacke** [comment]
Reminder for ordinal regression: https://twitter.com/ChelseaParlett/status/1352367383243939841?s=19

**Comment by DominiqueMakowski** [comment]
https://github.com/easystats/bayestestR/issues/20

**Comment by mattansb** [comment]
It is sigma!

![image](https://user-images.githubusercontent.com/35330040/105633579-09d9b280-5e62-11eb-9629-84df7d4c6eaf.png)

So... "fix" `rope_range()` with a startup warning?...

**Comment by mattansb** [comment]
It seems that he *also* gives the `0.1*sd(y)` as *another* option?

![image](https://user-images.githubusercontent.com/35330040/105634580-4e1b8180-5e67-11eb-8c70-0c92d99f73c2.png)


How very weird - these two options give very different results!

**Comment by DominiqueMakowski** [comment]
From my empirical experience, I felt like the current default rope range for logistic was "a bit larger" than for lms (i.e., it was a bit harder to be outside the rope given similarish effects). Is the new value usually bigger or smaller than the old?

**Comment by mattansb** [comment]
### 1. The conditional sd will be smaller than the unconditional sd.

So for the logistic case, the default we use is expected to be rather small - a change of of `0.1*3/sqrt(pi) = 0.18` ~ `OR: 1.2` is indeed [rather small](https://easystats.github.io/effectsize/articles/interpret.html#odds-ratio).

---


### 2. We have an inconsistency in the default rope range between models:

For linear models we use the unconditional, and for logistic we use conditional:
|    | Unconditional | Conditional |
| --- | --- | --- | 
| **Logistic** (any `link = "logit"`) | --none-- | `3/sqrt(pi)`
| **Gaussian** (really any `link = "identity"`) | S<sub>y</sub> = `sd(insight::get_response(model))` | `sigma(model)`

**Comment by DominiqueMakowski** [comment]
``` r
get_results <- function(n=100, d=0.5){
  df <- effectsize::standardize(bayestestR::simulate_difference(n=n, d=d))
  
  m1 <- glm(V0 ~ V1, data=df, family="binomial")
  m2 <- lm(V1 ~ V0, data=df)
                    
  data.frame(n=c(n, n, n),
             d=c(d, d, d), 
             type=c("LM", "GLM - basic", "GLM - sigma"),
             rope=c(bayestestR::rope_range(m2)[2], 0.1 * pi / sqrt(3), 0.1 * insight::get_sigma(m1)),
             coef=insight::get_parameters(m2)[2, 2], insight::get_parameters(m1)[2, 2], insight::get_parameters(m1)[2, 2])
}

data <- data.frame()
for(d in seq(0, 1, length.out = 100)){
  data <- rbind(data, get_results(d=d))
}

library(ggplot2)

ggplot(data, aes(x=d, y=rope, color=type)) +
  geom_point()
```

![](https://i.imgur.com/zRACnFG.png)

<sup>Created on 2021-01-26 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

Indeed, using sigma would result in mostly smaller ROPEs it seems. But I still have troubles with the fact that it depends on the number of 1s and 0s in the dependent variable when the logodds obtained from logistic models are supposed to be unrelated to such properties of the outcome (in their interpretation). I might be missing something out, though

If we have enough evidence to replace the current fixed value by 0.1 * sigma then okay.

Let's first finally submit bayestestR as is, and then for next version we address this + finally the CI range 😬

**Comment by DominiqueMakowski** [comment]
we can add a warning for the rope_range of logistic models

**Comment by mattansb** [comment]
@DominiqueMakowski not 100% sure what you're simulation was doing, but you're (again 😂) using the OR to d conversation wrong - it's not for predicting group from y in a logistic regression, it's for predicting a dichotomized y from group in a logistic regression.

Also, the warnings shouldn't be for logistic models (as I noted, there is no unconditional alternative for `3/sqrt(pi)`), but for **linear** models, where we might change from the current unconditional to the conditional.

**Comment by mattansb** [comment]
This is what I'm trying to say:

``` r
get_results <- function(n=100, d=0.5){
  df <- bayestestR::simulate_difference(n=n, d=d)
  
  m1 <- glm(I(V1>median(V1)) ~ V0, data=df, family="binomial")
  m2 <- lm(V1 ~ V0, data=df)
  
  data.frame(n=c(n, n, n),
             d=c(d, d, d), 
             type=c("LM - current (S)", "LM - sigma", "GLM (rescaled)"),
             rope=c(bayestestR::rope_range(m2)[2], 0.1 * insight::get_sigma(m2), 0.1), # rescaled, 0.1 * pi / sqrt(3) is just 0.1
             coef=c(insight::get_parameters(m2)[2, 2], insight::get_parameters(m2)[2, 2], insight::get_parameters(m1)[2, 2]))
}

library(purrr)
data <- map_dfr(seq(0, 2, length.out = 50), get_results, n = 100)


library(ggplot2)

ggplot(data, aes(x=d, y=rope, color=type)) +
  geom_line(size = 2) +
  coord_cartesian(ylim = c(0.05,0.15))
```

![](https://i.imgur.com/3Dl3ygQ.png)

<sup>Created on 2021-01-26 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by DominiqueMakowski** [comment]
aaaaaah I entirely misread this thread then

**Comment by mattansb** [comment]
lol

**Comment by DominiqueMakowski** [comment]
My priors made me think it was a rehash of #20, my deepest apologies 😅

**Comment by mattansb** [comment]
Classic example of bad priors :P

**Comment by mattansb** [comment]
I cleaned-up the code a bit, and put the warning in the right place.
We can finish resolving before the *next* CRAN release.

**Comment by DominiqueMakowski** [comment]
re-submitting, then

**Comment by DominiqueMakowski** [comment]
resubmitted ^^

**Comment by DominiqueMakowski** [comment]
Dear maintainer,

package bayestestR_0.8.2.tar.gz does not pass the incoming checks automatically, please see the following pre-tests:
Windows: <https://win-builder.r-project.org/incoming_pretest/bayestestR_0.8.2_20210126_062500/Windows/00check.log>
Status: OK
Debian: <https://win-builder.r-project.org/incoming_pretest/bayestestR_0.8.2_20210126_062500/Debian/00check.log>
Status: 1 NOTE

**Comment by mattansb** [comment]
Because of the note?? The URL works fine...

**Comment by strengejacke** [comment]
Do you have an attached text file that also contains another check log?

**Comment by DominiqueMakowski** [comment]
Lol dont worry I just pasted that here for reference but I'll look into it as soon as I reach soon :)

**Comment by DominiqueMakowski** [comment]
I'll just reply say it's prolly a false positive :)

**Comment by strengejacke** [comment]
Or remove link, merge https://github.com/easystats/bayestestR/pull/367, and re-submit 8-)

**Comment by DominiqueMakowski** [comment]
good idea, though I already replied, I'll resubmit nonetheless

**Comment by DominiqueMakowski** [comment]
re-re-submitted 🤞

**Comment by DominiqueMakowski** [comment]
On CRAN

**Comment by bbolker** [comment]
help ... it's great that this warning exists (I guess), but it's super-opaque to figure out what to do about it if it pops up (apparently) out of nowhere ...

I'm getting it from running `dotwhisker::dwplot()` on a `stanreg` object. `dw_tidy` is calling `broomExtra::tidy_parameters` which is eventually making its way down the chain to `rope`. The message *says* it's about binomial models but is actually about linear models ...?

reprex:

```
library(rstanarm)
library(dotwhisker)
fit <- stan_lm(mpg ~ wt + qsec + cyl, data = mtcars, prior = R2(0.75),
               refresh=0, seed=101)
dwplot(fit)
```

Session info has 111 packages (if you want I can post the whole thing) but this is dotwhisker-0.6.0, rstanarm 2.21.1, bayestestR 0.8.2

> Warning in doTryCatch(return(expr), name, parentenv, handler): Note that the
## default rope range for binomial models might change in future versions (see
## https://github.com/easystats/bayestestR/issues/364).Please set it explicitly to
## preserve current results.

```
where 1: rope(x, range = rope_range, ci = rope_ci, ...)
where 2: .describe_posterior(posteriors, centrality = centrality, dispersion = dispersion, 
    ci = ci, ci_method = ci_method, test = test, rope_range = rope_range, 
    rope_ci = rope_ci, bf_prior = bf_prior, BF = BF, effects = effects, 
    component = component, parameters = parameters, ...)
where 3: describe_posterior.stanreg(model, centrality = centrality, dispersion = dispersion, 
    ci = ci, ci_method = ci_method, test = test, rope_range = rope_range, 
    rope_ci = rope_ci, bf_prior = bf_prior, diagnostic = diagnostic, 
    priors = priors, ...)
where 4: bayestestR::describe_posterior(model, centrality = centrality, 
    dispersion = dispersion, ci = ci, ci_method = ci_method, 
    test = test, rope_range = rope_range, rope_ci = rope_ci, 
    bf_prior = bf_prior, diagnostic = diagnostic, priors = priors, 
    ...)
where 5: .extract_parameters_bayesian(model, centrality = centrality, 
    dispersion = dispersion, ci = ci, ci_method = ci_method, 
    test = test, rope_range = rope_range, rope_ci = rope_ci, 
    bf_prior = bf_prior, diagnostic = diagnostic, priors = priors, 
    effects = effects, standardize = standardize, ...)
where 6: model_parameters.stanreg(x, ...)
where 7: parameters::model_parameters(x, ...)
where 8: standardize_names(parameters::model_parameters(x, ...), style = "broom")
where 9: doTryCatch(return(expr), name, parentenv, handler)
where 10: tryCatchOne(expr, names, parentenv, handlers[[1L]])
where 11: tryCatchList(expr, classes, parentenv, handlers)
where 12: tryCatch(expr = standardize_names(parameters::model_parameters(x, 
    ...), style = "broom"), error = function(e) NULL)
where 13: broomExtra::tidy_parameters(x, conf.int = TRUE, ...)
where 14: dw_tidy(x, ci, by_2sd, margins, ...)
where 15: dwplot(lms)
```

**Comment by strengejacke** [comment]
> The message says it's about binomial models but is actually about linear models ...?

Yes, this was a misnomer fixed in this commit, but not yet in the CRAN version:

https://github.com/easystats/bayestestR/commit/e70ec3eac0571f879b1691f910ce33071d971ded#diff-8a9dfe10a23f01c6067d481572ad8f67365347367637e87548e6fa48f1119d54

You can use `verbose` argument, or better, @IndrajeetPatil do you pass down verbose to rope_range?

**Comment by mattansb** [comment]
I think it would perhaps be better to have this in the startup message, what do you think?

Also, @bbolker while you're here... Care to chime in? 😅 Should the default rope be based on `sd(y)` or `sigma`?

**Comment by strengejacke** [comment]
> I think it would perhaps be better to have this in the startup message, what do you think?

Or we resolve this issue and can then remove the message ;-)

**Comment by strengejacke** [comment]
The question is whether the conditional "rope" makes more sense than the unconditional one... The idea behind the ROPE is to have a relationship/association/effectsize between predictors and outcome that can be considered as "large enough" to be relevant. From my understanding, "large enough" refers to a significant _change in the outcome_ - i.e. the unconditional rope range (+/- 0.1 * SD(y)).

**Comment by DominiqueMakowski** [comment]
> From my understanding, "large enough" refers to a significant change in the outcome - i.e. the unconditional rope range (+/- 0.1 * SD(y))

I spontaneously tend to agree

**Comment by strengejacke** [comment]
One could argue that, due to multiple regression, the "raw" effect of a predictor (i.e. unconditional rope range) is not of interest, since coefficients are "adjusted" for each other. That could mean we should use the _conditional_ rope range. However, in real worlds, we always have confounders and no "isolated" associations, but we _still_ want to have a relevant change in our outcome - even if conditioned on / adjusted for other covariates. Then we would still look for +/- 0.1 * SD(y) - unless there's a good reason that the unconditional rope range is biased, and the conditional rope range is preferred.

**Comment by IndrajeetPatil** [comment]
> You can use verbose argument, or better, @IndrajeetPatil do you pass down verbose to rope_range?

Not explicitly. But everything in `...` given to `tidy_parameters` should pass to `model_parameters`.

``` r
library(rstanarm)
#> Loading required package: Rcpp
#> This is rstanarm version 2.21.1
#> - See https://mc-stan.org/rstanarm/articles/priors for changes to default priors!
#> - Default priors may change, so it's safest to specify priors, even if equivalent to the defaults.
#> - For execution on a local, multicore CPU with excess RAM we recommend calling
#>   options(mc.cores = parallel::detectCores())
library(dotwhisker)
#> Loading required package: ggplot2
#> Registered S3 method overwritten by 'broom.mixed':
#>   method      from 
#>   tidy.gamlss broom

fit <- stan_lm(mpg ~ wt + qsec + cyl,
  data = mtcars, prior = R2(0.75),
  refresh = 0, seed = 101
)
#> Warning: There were 12 divergent transitions after warmup. See
#> http://mc-stan.org/misc/warnings.html#divergent-transitions-after-warmup
#> to find out why this is a problem and how to eliminate them.
#> Warning: Examine the pairs() plot to diagnose sampling problems

dwplot(fit, verbose = FALSE)
#> Warning: Note that the default rope range for linear models might change in
#> future versions (see https://github.com/easystats/bayestestR/issues/364).Please
#> set it explicitly to preserve current results.
#> Possible multicollinearity between cyl and wt (r = 0.84), cyl and qsec (r = 0.71). This might lead to inappropriate results. See 'Details' in '?rope'.
#> Warning: Note that the default rope range for linear models might change in
#> future versions (see https://github.com/easystats/bayestestR/issues/364).Please
#> set it explicitly to preserve current results.
#> Possible multicollinearity between cyl and wt (r = 0.84), cyl and qsec (r = 0.71). This might lead to inappropriate results. See 'Details' in '?rope'.
```

![](https://i.imgur.com/RAIusiU.png)

<sup>Created on 2021-03-23 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

**Comment by strengejacke** [comment]
The verbose-argument wasn't properly passed to `describe_posterior()` from `model_parameters()`, but this is fixed now:

``` r
library(rstanarm)
#> Loading required package: Rcpp
#> This is rstanarm version 2.21.1
#> - See https://mc-stan.org/rstanarm/articles/priors for changes to default priors!
#> - Default priors may change, so it's safest to specify priors, even if equivalent to the defaults.
#> - For execution on a local, multicore CPU with excess RAM we recommend calling
#>   options(mc.cores = parallel::detectCores())
library(dotwhisker)
#> Loading required package: ggplot2
#> Registered S3 method overwritten by 'broom.mixed':
#>   method      from 
#>   tidy.gamlss broom

fit <- stan_lm(mpg ~ wt + qsec + cyl,
               data = mtcars, prior = R2(0.75),
               refresh = 0, seed = 101
)
#> Warning: There were 19 divergent transitions after warmup. See
#> http://mc-stan.org/misc/warnings.html#divergent-transitions-after-warmup
#> to find out why this is a problem and how to eliminate them.
#> Warning: Examine the pairs() plot to diagnose sampling problems
dwplot(fit, verbose = FALSE)
```

![](https://i.imgur.com/HgSeIkK.png)

<sup>Created on 2021-03-23 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

**Comment by mattansb** [comment]
I think there are reasons for each of the options (and as you know I think that defaults should be avoided anyway)... But,

My main concern (which is how this whole thing started https://github.com/easystats/bayestestR/issues/364#issue-791943041) is that for GLMs, we return the "conditional"-based rope, and so there is an issue of **consistency** if we return the unconditional-rope only for gaussian models.

**Comment by strengejacke** [comment]
We could keep this inconsistency top secret, delete this issue and, to be sure, the whole internet as well... 😬

**Comment by DominiqueMakowski** [comment]
should we have it as an option to toggle? Fixed vs. Conditional? In the first case, it would be 0.1 * SD(y) and the equivalent based on the log-odds -> transformation for logit, and in the second we base it on sigma?

**Comment by DominiqueMakowski** [comment]
The above seems good!

**Comment by DominiqueMakowski** [comment]
|    | Marginal | Conditional / Default |
| --- | --- | --- | 
| **Logistic** (any `link = "logit"`) | *NA* | `3/sqrt(pi)`
| **Gaussian** (any `link = "identity"`) | S<sub>y</sub> = `sd(insight::get_response(model))` | `sigma(model)`

**Comment by mattansb** [comment]
Yes. So we should add a note in the startup message that the default rope for linear models is changed.

**Comment by DominiqueMakowski** [comment]
but currently the default is `sd(insight::get_response(model))` so nothing will change?

**Comment by mattansb** [comment]
Ah, you switched [my table](https://github.com/easystats/bayestestR/issues/364#issuecomment-766621744) around!

Should be this:

|  Model Type  | Marginal | Conditional (default) |
| --- | --- | --- | 
| **Logistic** (any `link = "logit"`) | *NA* | `3/sqrt(pi)`
| **Gaussian** (any `link = "identity"`) | S<sub>y</sub> = `sd(insight::get_response(model))` | `sigma(model)` |

**Comment by DominiqueMakowski** [comment]
Oops my bad. Something that will need to be explained and clarified that might bug people's intuition is that for lms, the better the model the lower sigma and the smaller the rope, but for glms the rope doesn't change depending on the model quality, it's always the same value

**Comment by mattansb** [comment]
That's true - for the same data.

However, for different samples, over different ranges of Xs (say you sample everyone and I only sample children) sigma is (or should be) the same.

I can add a note to the relevant vignette, that you can reference in the docs and startup message. Sounds good?

**Comment by DominiqueMakowski** [comment]
> I can add a note to the relevant vignette, that you can reference in the docs and startup message. Sounds good?

Yeah

But wouldn't 3/sqrt(pi) also apply for the marginal ROPE (for logit)? I'm trying to wrap my head around the implications

**Comment by mattansb** [comment]
Nope... GLMs don't have a straight forward way of getting the marginal variance of the latent variable 🤷‍♂️

**Comment by DominiqueMakowski** [comment]
> We could keep this inconsistency top secret, delete this issue and, to be sure, the whole internet as well... 😬

I'm starting to think that this is actually a not-so-bad alternative 😁

**Comment by DominiqueMakowski** [comment]
To be honest, while I understand the crux of this issue (though I still have trouble understanding why 3/sqrt(pi) is *conditional* and why is it fundamentally different from the marginal linear version), the current behaviour (i.e., "marginal" for linear and "conditional" for logit) is really straightforward to explain and transparent to implement

We discuss(ed) it clearly in the docs, posts and papers, so it's really weird that nobody else had seemed to notice the discrepancy (especially if it leads to some problems)... I mean the Bayesian framework have been under so much scrutiny lately, and ROPEs are even used outside of that context (btw I wonder what would be Lacken's perspective on that)... 

This relative absence of agitation surrounding our previous default suggests that it's not a timely issue: we can probablyt run a bit more with it 

So I'll suggest adding a `method = "legacy"` argument, which corresponds to our original (inconsistent) default. We can explain in the docs a bit about the issue and say that additional methods will be implemented in the future (and we link to that discussion). Then, we <sub>~"delete this issue and, to be sure, the whole internet as well"~</sub> let this issue sit around for some more time until Mat' has a eureka moment in his bath, realizing that he's been wrong all along 😁 Or until we gather some more thoughts on this

**Comment by mattansb** [comment]
Yes, I also find it weird that no one has pointed this out. However, the might be because (1) <sub><sub>serious</sub></sub> people don't use the default values, or (2) they tend towards BFs more than ROPEs?

I agree that we can let this steam some more - no need to add the `method` argument at all right now, I think.

---

> though I still have trouble understanding why ~3/sqrt(pi)~ pi / sqrt(3) is conditional and why is it fundamentally different from the marginal linear version

This is the definition of a logistic regression:

`P(y==1) = 1/(exp(-η))`, where `η` is a latent variable (we only see it's dichotomized binary form) that is conditionally normally distributed ::: `η ~ N(bX, pi / sqrt(3))` (where `bX` is the linear part we usually care about).

This is also why `log(OR) * sqrt(3) / pi` is "equivalent" to Cohen's d:

- `log(OR)` is the distance between two groups, standardized by the conditional variance `pi / sqrt(3)` (on the latent scale).
- `M1-M2` is the distance between two groups, standardized by the conditional variance `sigma`.

---

Here's some more complexity:

In a (G)LMM, the default rope range should probably be conditioned on the level of the effect.
1. Level 1 effects `sigma` - a within-person effect should be compared to the variance within people, no?
2. Level 2 effects `Var(intercept)` - a between-person effect should be compared to the variance between people, no?
3. ... 

---

## And remember kids - don't use defaults!

**Comment by DominiqueMakowski** [comment]
thanks for the explanations, (and your patience), it's much clearer now ☺️
